### PR TITLE
feat: BigQuery Faster Sync

### LINKED ISSUE #48094: Faster sync on BigQuery
Apply the same optimizations we did for Redshfit on BigQuery
https://github.com/metabase/metabase/issues/39986

[Comparison dashboard](https://stats.metabase.com/dashboard/969-task-performance-all-databases-cloud?database_age_%28in_months%29_greater_than_or_equal_to=1&database_creation_date=&databases=bigquery-cloud-sdk&databases=redshift&date_filter=past12months~)

### PR DESCRIPTION
**Pull request by snoe**
Replaces the O(n-tables) `describe database` and `describe-tables` with O(datasets) versions that query `INFORMATION_SCHEMA`

Fixes #48094

### CODE REVIEW

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj [hunk 1]
```diff
@@ -291,39 +300,104 @@
...
+                                 :table-name table-name
+                                 :table-schema dataset-id
+                                 :database-type database-type
+                                 :base-type base-type
+                                 :nfc-path nfc-path}))
+                            results)]
+    (reduce
+     (fn [accum col]
+       (when-let [parent (:nfc-path col)]
+         (update-in accum [(:table-name col) parent] (fnil conj []) col)))
```
Comment (metamben) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:328]: I expected a `let` and a `cond-> accum parent ...` combination. Is there a test covering this?

File: modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj [hunk 2]
```diff
@@ -705,12 +777,10 @@
...
                       (fn [tbl-nm] ["DROP TABLE IF EXISTS `%s.%s`" test-db-name tbl-nm])
                       (fn [tbl-nm]
-                        (is (= {:schema test-db-name
-                                :name   tbl-nm
-                                :fields #{{:name "int_col" :database-type "INTEGER" :base-type :type/Integer :database-position 0 :database-partitioned false}
-                                          {:name "array_col" :database-type "INTEGER" :base-type :type/Array :database-position 1 :database-partitioned false}}}
-                               (driver/describe-table :bigquery-cloud-sdk (mt/db) {:name tbl-nm :schema test-db-name}))
-                            "`describe-table` should detect the correct base-type for array type columns")))))
+                        (is (= [{:name "int_col" :database-type "INTEGER" :base-type :type/Integer :database-position 0 :database-partitioned false :table-name tbl-nm :table-schema test-db-name}
+                                {:name "array_col" :database-type "ARRAY" :base-type :type/Array :database-position 1 :database-partitioned false :table-name tbl-nm :table-schema test-db-name}]
```
Comment (metamben) [inline modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj:781]: Do you know that switching from "INTEGER" to "ARRAY" is safe? Or do we trust existing tests to tell us if this is OK?

**Review by metamben — APPROVED** [review]
Should be good when all tests pass.

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj [hunk 3]
```diff
@@ -291,39 +300,104 @@
...
+  (let [results (query-honeysql
+                 driver
+                 database
+                 (cond->
+                  {:select [:table_name :column_name :data_type :field_path]
+                   :from [[(information-schema-table project-id dataset-id "COLUMN_FIELD_PATHS") :c]]}
+                   (not-empty table-names)
+                   (assoc :where [:in :table_name table-names])))
+        nested-columns (map (fn [{data-type :data_type field-path-str :field_path table-name :table_name}]
+                              (let [field-path (str/split field-path-str #"\.")
```
[the remark this answers is not in the document]
Comment (snoe) [reply modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:318]: They can't https://cloud.google.com/bigquery/docs/schemas#column_names (also explicity excluded from "flexible" column names)

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj (hunk 1 above)
Comment (snoe) [reply modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj]: Good catch I don't know how `sync-nested-fields-test` missed it.

**Pull request merged**